### ISSUE TITLE
Plugin filled its own directory with >3500 temp files

### ISSUE THREAD

**Issue by LPGhatguy** [Microsoft/TypeScript-Sublime-Plugin#348]
I opened the plugin folder today to update the `tsserver` and found that there were 3686 (!) temp files of the form `.tmpbufXXXX` in the folder. They total to about 20 MB.

What is their purpose? Are these supposed to be cleaned out periodically?

**Comment by hoanhtien** [comment]
To reload the content of a file in tsserver, we need a way to pass the new file content to the server. The current way of doing it is to store the file content to a temp file and send the temp file location to the server.

Those temp files are designed to be recyclable. We only expect a small number of them to be created. Note that we only need to 'reload' the file that we are currently editing.

The problem here could be that after receiving a 'reload' command, by some reason (e.g. a bug) tsserver fails to send back an ACK response, the temp file used in this command will not be marked to for recycling. If this problem presents for a while, new temp files will be created and start building up.

Those temp files can be safely deleted after existing Sublime Text. We can add a logic to clean them during starting/closing Sublime Text.

**Comment by zhengbli** [comment]
The number of file should be so large though. How large is your project?

**Comment by LPGhatguy** [comment]
The project isn't too large: It has 52 core code files and 7 external typed dependencies (which have large `.d.ts` files)

**Comment by zhengbli** [comment]
The tmpbuf files are only used to reload the buffer. So the supposed process is:
1. the plugin writes the buffer content to a tmpbuf file
2. the plugin sends a request to the server to load from that tmpbuf file
3. the server sends back a response after the loading is finished
4. upon receiving the server's response, the plugin marked the tmpbuf file as ready to reuse

Therefore the most likely explanation of having so many tmpbuf files is that the server didn't respond after finishing the reload. So could you do the following to help us diagnose the problem:
1. go to the "[Plugin Folder]\typescript\libs\global_vars.py", change the ``LOG_CONSOLE_LEVEL = logging.DEBUG``
2. restart sublime, and try to repro the problem
3. open your sublime console via ``ctrl + ` ``, paste the log info here

**Comment by zhengbli** [comment]
@LPGhatguy And as this issue does have some serious perf impact, would you be able to do a skype session if that would us find the problem quicker? I tried to edit files in your https://github.com/LPGhatguy/basis repo, didn't notice much differences. Thanks

**Comment by LPGhatguy** [comment]
Yeah, definitely. Shoot me an email and we can plan out when to do that.

**Comment by lorem--ipsum** [comment]
I have a similar issue, where the directory is filled with more than 6000 `tmpbuf` files. I'd be happy to help solving this issue !

**Comment by jonschlinkert** [comment]
> We can add a logic to clean them during starting/closing Sublime Text.

I think this is necessary. There are more than 80,000 of these files in my folder.

![image](https://user-images.githubusercontent.com/383994/63215822-1b118200-c0fa-11e9-99b4-8c313e378c55.png)

My recommendation would be something like the following: 

- Put all of these folders in a runtime cache folder in your plugin's directory, like `/TypeScript/cache`
- Add logic to delete the `cache` folder when sublime text exits. This should probably be done in a separate process that periodically checks to see if the `plugin_host` still exists (e.g. the parent process).
- Alternatively, you could just delete the `cache` on startup instead of exit.
